### ISSUE TITLE
Vault.ssl_ca_cert method no longer works

### ISSUE THREAD

**Issue by mfischer-zd** [hashicorp/vault-ruby#108]
As of v0.7.0, the following code no longer works:

```
Vault.ssl_ca_cert = '/path/to/cert'
```

As a consequence, requests are now failing with certificate verification errors.

Interestingly, passing the certificate path to `Vault::Client.new(ssl_ca_cert: '/path/to/cert')` does work as before.

**Comment by jfreimuth** [comment]
As of v0.7.0, the following code no longer works:

`Vault.auth.tls(pem)`

As a consequence, requests are now failing with

"The Vault server at `https://vault-develop-ci.lhotse.ov.otto.de' is not currently
accepting connections. Please ensure that the server is running and that your
authentication information is correct.

The original error was `Net::HTTP::Persistent::Error'. Additional information (if any) is
shown below:

    too many connection resets (due to end of file reached - EOFError) after 0 requests on 70312685070080, last used 1476947703.442755 seconds ago

Please refer to the documentation for more help."

**Comment by jfreimuth** [comment]
Same Problem with

Vault.ssl_pem_contents

**Comment by grosser** [comment]
@evanphx

**Comment by sethvargo** [comment]
Hi all,

We triage issues routinely. There's no need to CC folks on issues and PRs - they will review them when they have time. Thanks! 😄

**Comment by evanphx** [comment]
@mfischer-zd In the case where you're doing `Vault.ssl_ca_cert`, how are you making requests? Is it by doing `Vault.get` or are you using `Vault::Client.new.get`?

The reason being the code was changed around July 2015 such that the settings set directly on `Vault` are only used by the `Vault.client` instance, not any other created instances.

**Comment by mfischer-zd** [comment]
The code that began failing is, roughly:

```ruby
Vault.ssl_ca_cert = ca_cert_path

Vault.with_retries(Vault::HTTPConnectionError, attempts: new_resource.attempts) do |attempt, e|
  request_params = {
    common_name: cn
  }
  response = Vault.logical.write(request_uri, request_params)
  key = response.data[:private_key]
  cert = response.data[:certificate]
  ca = response.data[:issuing_ca]
end
```

**Comment by BRMatt** [comment]
I've spent the past hour trying to track down a related bug, I think the cause is related to the introduction of persistent connections. When vault is required [it sets up a singleton `Vault.client`](https://github.com/hashicorp/vault-ruby/blob/fa059fd605695bc8cfc5b7f41b9ddb4c0a0484f7/lib/vault.rb#L18). Calling `Vault.ssl_ciphers=`/`Vault.ssl_ca_cert=` sets configuration [on this client instance](https://github.com/hashicorp/vault-ruby/blob/fa059fd605695bc8cfc5b7f41b9ddb4c0a0484f7/lib/vault.rb#L35).

In `0.6.0` a new http connection would be created from the client's configuration on every request, but as of 0.7.0, the HTTP connection is ["created" and configured in `Vault::Client`'s initializer](https://github.com/hashicorp/vault-ruby/blob/fa059fd605695bc8cfc5b7f41b9ddb4c0a0484f7/lib/vault/client.rb#L78-L141), which means any configuration changes are essentially a no-op, as they're never passed through to the underlying HTTP connection.

**Comment by mfischer-zd** [comment]
@BRMatt that sounds like a different problem, though certainly one worth resolving.  Would you mind filing another issue to track that separately?

**Comment by BRMatt** [comment]
@mfischer-zd Really? It seems related to me? Changing the SSL options of the singleton `Vault.client` via `Vault.XXX = `doesn't have any effect, but creating a new Vault client and passing the options directly to it cause the options to take effect? Isn't that the problem you're having?

**Comment by evanphx** [comment]
@BRMatt That is totally an issue the new persistent code. I'll get that fixed ASAP.

**Comment by ferrous26** [comment]
I'm experiencing a very similar problem with `ssl_ca_path`. The standard vault go client is happy with the configuration, but this ruby implementation seems to fail on validating the server certificate.

In my case, making a separate client does not help.

ruby 2.4.1, vault-ruby 0.10.1

**Comment by abelcastilloavant** [comment]
Possibly related: I was able to get vault-ruby to find my certificate by specifying its location with the env var `VAULT_CA_CERT`. It fails to find the certificate when I specify its location with the env var `VAULT_SSL_CERT`.

ruby >2.3, vault-ruby 0.11.0

**Comment by JustinAddams** [comment]
Unsure if this information helps or is related, but when configuring Vault.ssl_ca_cert = 'CRT FILE LOCATION' it fails with the above mentioned errors unless I specifically add the cert chain to the trust store.

This might be common knowledge, I am unsure. But it was something not mentioned in any doco I read.

Version: ruby 2.4.1, vault-ruby 0.11.0

**Comment by JustinAddams** [comment]
Ignore my previous comment, user error.
Using `Vault.ssl_ca_cert = '/home/user/ca.crt'` is working fine for me.

Please note the .crt file I am providing is the "Full Chain" and internally created.
ruby 2.4.1, vault-ruby 0.11.0

**Comment by evanphx** [comment]
Sounds like this is fixed. Please open a new issue if you have the problem again.

**Comment by rajsahae** [comment]
I know this issue is closed but with vault-ruby 0.12.0 I'm having this exact issue. Setting Vault.ssl_ca_cert doesn't work when using the singleton but if I instantiate a client and set it on the client, then it works fine.
